### ISSUE TITLE
LinkifyFilter with clean breaks links at &amp;

### ISSUE THREAD

**Issue by Deimos** [mozilla/bleach#422]
(Split out to a new issue from [this comment](https://github.com/mozilla/bleach/issues/419#issuecomment-445056894))

When using `LinkifyFilter` instead of the stand-alone `linkify`, links are broken at any `&amp;`, but not bare `&` characters.

Here's a pretty minimal demonstration:

```
import bleach

bleach.linkify("https://www.youtube.com/watch?v=SkcucKDrbOI&amp;feature=youtu.be")
```

Result (expected): `<a href="https://www.youtube.com/watch?v=SkcucKDrbOI&amp;feature=youtu.be" rel="nofollow">https://www.youtube.com/watch?v=SkcucKDrbOI&amp;feature=youtu.be</a>`

```
import bleach
from functools import partial

bleach_linkifier = partial(bleach.linkifier.LinkifyFilter)
cleaner = bleach.Cleaner(filters=[bleach_linkifier])
cleaner.clean("https://www.youtube.com/watch?v=SkcucKDrbOI&amp;feature=youtu.be")
```

Result (not expected, link stops at the `&amp;`): `<a href="https://www.youtube.com/watch?v=SkcucKDrbOI">https://www.youtube.com/watch?v=SkcucKDrbOI</a>&amp;feature=<a href="http://youtu.be">youtu.be</a>`

It works with just a bare `&` instead of the `&amp;`, like `https://www.youtube.com/watch?v=SkcucKDrbOI&feature=youtu.be`. In that case, the output of both methods is:

```
<a href="https://www.youtube.com/watch?v=SkcucKDrbOI&amp;feature=youtu.be">https://www.youtube.com/watch?v=SkcucKDrbOI&amp;feature=youtu.be</a>
```

(with an additional `rel="nofollow"` attr added by `linkify()` by default)

**Comment by Deimos** [comment]
Sorry if this is obvious/irrelevant, but would it be related to `consume_entities` being false for `Cleaner` but true for `Linker`?

**Comment by willkg** [comment]
Yes! And no-ish. I added the `consume_entities` flag because that was the only way I could fix #392 by letting linkify use the `BleachHTMLParser` and friends. So if you use `linkify`, then the tokenizer consumes entities and the `&amp;` becomes an `&` and when `LinkifyFilter` sees it, it's one big `Characters` token and it linkifies the whole url, and then the serializer escapes the `&` converting it back to an `&amp;`. So `linkify` on its own works.

The problem is that when using the `LinkifyFilter` with the cleaner, we don't want the tokenizer to consume entities because that lossily transforms the fragment and then encoding/decoding gets hard to reason about and lossy and some of the entities disappear. There's a whole discussion about that in one of the PRs. So I added the `BleachHTMLTokenizer` which leaves entities as they are and they flow through just fine except in the case where you're using a `LinkifyFilter` and the url you want linkified contains an entity which causes the url to be broken across multiple tokens.

I think I may add some peeking to the stream and just handle the `&amp;` case. `&amp;` is really common in urls. I don't think other entities are as common. If they are, we can special case them, too.

I'll fiddle with it this week.
